File: src/main/java/com/grizzly/baseViews/AbstractConstraintLayout.java
package com.grizzly.baseViews;

import android.content.Context;
import android.util.AttributeSet;

/**
 * Created by FcoPardo on 3/6/16.
 */
public abstract class AbstractConstraintLayout<T> extends BaseConstraintLayout implements BaseView.OnDataDrivenView<T> {

    protected T data;
    protected Class<T> dataClass;

    public AbstractConstraintLayout(Context context) {
        super(context);
    }

    public AbstractConstraintLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public AbstractConstraintLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    protected abstract void setControls();

    public <X extends BaseView.OnDataDrivenView<T>> X setData(T aData) {
        this.data = aData;
        if(data != null)setControls();
        return (X) this;
    }

    public T getData() {
        return data;
    }
}

File: src/main/java/com/grizzly/baseViews/AbstractConstraintLayout.kt
package com.grizzly.baseViews

import android.content.Context
import android.util.AttributeSet

abstract class AbstractConstraintLayout<T> : BaseConstraintLayout, BaseView.OnDataDrivenView<T> {

    constructor(context : Context) : super(context)
    constructor(context : Context, attrs : AttributeSet) : super(context, attrs)
    constructor(context : Context, attrs : AttributeSet, styleAttr : Int) : super(context, attrs, styleAttr)

    @JvmField
    protected var myData: T? = null
    @JvmField
    protected var dataClass: Class<T>? = null

    override fun <X : BaseView.OnDataDrivenView<T>> setData(aData: T): X {
        this.myData = aData
        if (myData != null) setControls()
        return this as X
    }

    override fun getData(): T? {
        return myData
    }

    protected abstract fun setControls()
}